### PR TITLE
Raise fault on opflex agent if requested MTU exceeds fabric uplink MTU

### PR DESCRIPTION
**Pull request by Bhavanaashok33**
also added support to raise the fault if IP address not set for OpFlex link, not able to find suitable host uplink interface for vlan, not able to find parent link for OpFlex interface and Could not enumerate link addresses
Verified the changes on fabric if MTU exceeds fabric uplink MTU

### CODE REVIEW

**Review by tomflynn — CHANGES_REQUESTED** [review]

File: pkg/hostagent/opflex.go [hunk 1]
```diff
@@ -24,10 +24,55 @@ import (
 	"reflect"
 	"text/template"
 
+	"encoding/json"
+	gouuid "github.com/nu7hatch/gouuid"
```
[the remark this answers is not in the document]
Comment (Bhavanaashok33) [reply pkg/hostagent/opflex.go:28]: There is similar implementation on pkg/hostagent/pods.go where they have used nu7hatch/gouuid to generate Endpoint UUID. Thought of making use of this library. 
Sure, I will switch to github.com/google/uuid instead.

File: pkg/hostagent/opflex.go [hunk 2]
```diff
@@ -24,10 +24,55 @@ import (
...
 	"github.com/sirupsen/logrus"
 	"github.com/vishvananda/netlink"
 )
 
+type opflexFault struct {
+	FaultUUID   string `json:"fault_uuid"`
+	Severity    string `json:"severity"`
+	Description string `json:"description"`
+	FaultCode   string `json:"faultCode"`
```
[the remark this answers is not in the document]
Comment (Bhavanaashok33) [reply pkg/hostagent/opflex.go:37]: Yes, thats right

File: pkg/hostagent/opflex.go [hunk 3]
```diff
@@ -24,10 +24,55 @@ import (
...
+	uuid, _ := gouuid.NewV4()
+	FaultFilePath := filepath.Join(agent.config.OpFlexFaultDir, description+".fs")
+	faultFileExists := fileExists(FaultFilePath)
+	if faultFileExists {
+		agent.log.Debug("fault file exist")
+		return
+	}
+	fault := &opflexFault{
+		FaultUUID:   uuid.String(),
+		Severity:    "major",
```
[the remark this answers is not in the document]
Comment (Bhavanaashok33) [reply pkg/hostagent/opflex.go:63]: Sure will address this change

File: pkg/hostagent/opflex.go [hunk 4]
```diff
@@ -115,6 +170,8 @@ func (agent *HostAgent) discoverHostConfig() (conf *HostAgentNodeConfig) {
 					"name": link.Name,
 					"vlan": agent.config.AciInfraVlan,
 				}).Error("IP address not set for OpFlex link")
+				description := "IP_address_not_set_for_OpFlex_link"
+				agent.createMtuFault(description)
```
[the remark this answers is not in the document]
Comment (Bhavanaashok33) [reply pkg/hostagent/opflex.go:174]: Yup you are right. I am gonna change the method name from createMtuFault to createFaultOnAgent and will assign different fault code to each of them

File: pkg/hostagent/opflex.go [hunk 5]
```diff
@@ -22,12 +22,59 @@ import (
...
+	if err == nil && reflect.DeepEqual(existingdata, newdata) {
+		return false, nil
+	}
+	err = ioutil.WriteFile(faultfile, newdata, 0644)
+	return true, err
+}
+
+func (agent *HostAgent) createFaultOnAgent(description string, faultCode int) {
+	Uuid := uuid.New().String()
+	FaultFilePath := filepath.Join(agent.config.OpFlexFaultDir, description+".fs")
```
Comment (tomflynn) [inline pkg/hostagent/opflex.go:56]: leading lower case is convention for local variables I think

**Review by tomflynn — COMMENTED** [review]
Some minor nits. Please also take a look at the codacy complaints for this change at https://app.codacy.com/gh/noironetworks/aci-containers/pullRequest?prid=6757056#newIssuesView

File: pkg/hostagent/opflex.go (hunk 5 above)
Comment (Bhavanaashok33) [reply pkg/hostagent/opflex.go:56]: yes, will add this change

File: pkg/hostagent/opflex.go [hunk 6]
```diff
@@ -22,12 +22,59 @@ import (
...
+	desc := strings.Replace(description, "_", " ", -1)
+	fault := &opflexFault{
+		FaultUUID:   Uuid,
+		Severity:    "critical",
+		Description: desc,
+		FaultCode:   faultCode,
+	}
+	wrote, err := writeFault(FaultFilePath, fault)
+	if err != nil {
+		agent.log.Debug("Unable to write fault file")
```
[the remark this answers is not in the document]
Comment (Bhavanaashok33) [reply pkg/hostagent/opflex.go:71]: sure

File: pkg/hostagent/opflex.go [hunk 7]
```diff
@@ -22,12 +22,59 @@ import (
...
+	} else if wrote {
+		agent.log.Debug("Created fault file")
```
[the remark this answers is not in the document]
Comment (Bhavanaashok33) [reply pkg/hostagent/opflex.go:73]: okay

File: pkg/hostagent/agent.go [hunk 8]
```diff
@@ -385,7 +385,6 @@ func (agent *HostAgent) Run(stopCh <-chan struct{}) {
 		}
 		go agent.processSyncQueue(agent.syncQueue, stopCh)
 	}
-
```
Comment (tomflynn) [inline pkg/hostagent/agent.go:388]: I think this file doesn't need to be included in the changeset

**Review by tomflynn — COMMENTED** [review]
OK it looks good to me now but please add UTs that cover this code where possible and then we can merge

**Review by tomflynn — APPROVED** [review]

**Pull request merged**